fe/fe-core/src/main/java/com/starrocks/mysql/privilege/ImpersonateUserPrivTable.java

```diff
@@ -0,0 +1,91 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.mysql.privilege;
+
+import com.starrocks.analysis.UserIdentity;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.io.Text;
+import com.starrocks.persist.gson.GsonUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * ImpersonateUserPrivTable saves all impersonate user privileges
+ */
+public class ImpersonateUserPrivTable extends PrivTable {
+    private static final Logger LOG = LogManager.getLogger(ImpersonateUserPrivTable.class);
+
+    /**
+     * Return first priv which match the user@host on securedUser
+     * The returned priv will be saved in 'savedPrivs'.
+     **/
+    public void getPrivs(UserIdentity currentUser, UserIdentity securedUser, PrivBitSet savedPrivs) {
+        List<PrivEntry> userPrivEntryList = map.get(currentUser);
+        if (userPrivEntryList == null) {
+            return;
+        }
+
+        ImpersonateUserPrivEntry matchedEntry = null;
+        for (PrivEntry entry : userPrivEntryList) {
+            ImpersonateUserPrivEntry impersonateUserEntry = (ImpersonateUserPrivEntry) entry;
+
+            // check secured user
+            if (!impersonateUserEntry.getSecuredUserIdentity().equals(securedUser)) {
+                continue;
+            }
+
+            matchedEntry = impersonateUserEntry;
+            break;
+        }
+        if (matchedEntry == null) {
+            return;
+        }
+
+        savedPrivs.or(matchedEntry.getPrivSet());
+    }
+
+    /**
+     * check if currentUser can EXECUTE AS toUser
+     */
+    public boolean canImpersonate(UserIdentity currentUser, UserIdentity toUser) {
+        PrivBitSet privBitSet = new PrivBitSet();
+        getPrivs(currentUser, toUser, privBitSet);
+        return privBitSet.satisfy(PrivPredicate.IMPERSONATE);
+    }
+
+    public static ImpersonateUserPrivTable read(DataInput in) throws IOException {
```

Comment:
You should use GSON to serialize/deserialize this class. You can refer to DeleteHandler.

Reply:
Every entry of the table need to analyse() after deserialized. I use GSON to read every entry.

Reply:
You can process the analyse() after deserialize from GSON. If you want to add another persist property for this class, it will be impossible for this hard code serialize.

Reply:
done